p2p/transport/webrtc/udpmux/muxed_connection.go

```diff
@@ -0,0 +1,164 @@
+package udpmux
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"time"
+
+	pool "github.com/libp2p/go-buffer-pool"
+)
+
+var _ net.PacketConn = &muxedConnection{}
+
+const maxPacketsInQueue int = 128
+
+// muxedConnection provides a net.PacketConn abstraction
+// over packetQueue and adds the ability to store addresses
+// from which this connection (indexed by ufrag) received
+// data.
+type muxedConnection struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	pq     *packetQueue
+	// list of remote addresses associated with this connection.
+	// this is useful as a mapping from [address] -> ufrag
+	addresses []string
+	ufrag     string
+	mux       *udpMux
+}
+
+func newMuxedConnection(mux *udpMux, ufrag string) *muxedConnection {
+	ctx, cancel := context.WithCancel(mux.ctx)
+	return &muxedConnection{
+		ctx:    ctx,
+		cancel: cancel,
+		pq:     newPacketQueue(),
+		ufrag:  ufrag,
+		mux:    mux,
+	}
+}
+
+func (conn *muxedConnection) push(buf []byte, addr net.Addr) error {
+	return conn.pq.push(buf, addr)
+}
+
+// Close implements net.PacketConn
+func (conn *muxedConnection) Close() error {
+	_ = conn.closeConnection()
+	conn.mux.RemoveConnByUfrag(conn.ufrag)
+	return nil
+}
+
+// LocalAddr implements net.PacketConn
+func (conn *muxedConnection) LocalAddr() net.Addr {
+	return conn.mux.socket.LocalAddr()
+}
+
+// ReadFrom implements net.PacketConn
+func (conn *muxedConnection) ReadFrom(p []byte) (n int, addr net.Addr, err error) {
+	return conn.pq.pop(conn.ctx, p)
+}
+
+// SetDeadline implements net.PacketConn
+func (*muxedConnection) SetDeadline(t time.Time) error {
+	// no deadline is desired here
+	return nil
+}
+
+// SetReadDeadline implements net.PacketConn
+func (*muxedConnection) SetReadDeadline(t time.Time) error {
+	// no read deadline is desired here
+	return nil
+}
+
+// SetWriteDeadline implements net.PacketConn
+func (*muxedConnection) SetWriteDeadline(t time.Time) error {
+	// no write deadline is desired here
+	return nil
+}
+
+// WriteTo implements net.PacketConn
+func (conn *muxedConnection) WriteTo(p []byte, addr net.Addr) (n int, err error) {
+	return conn.mux.writeTo(p, addr)
+}
+
+func (conn *muxedConnection) closeConnection() error {
+	select {
+	case <-conn.ctx.Done():
+		return errAlreadyClosed
+	default:
+	}
+	conn.pq.close()
+	conn.cancel()
+	return nil
+}
+
+type packet struct {
+	addr net.Addr
+	buf  []byte
+}
+
+var (
+	errTooManyPackets = errors.New("too many packets in queue; dropping")
+)
+
+// just a convenience wrapper around a channel
+type packetQueue struct {
+	pkts   chan packet
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func newPacketQueue() *packetQueue {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &packetQueue{
+		ctx:    ctx,
+		cancel: cancel,
+		pkts:   make(chan packet, maxPacketsInQueue),
```

Comment:
This is quite a large buffer, consuming 1 kB, even if not in use. Any way we can refactor this to a `[]packet`?

Reply:
Going to think about it.
But that would make it not as easy, out-of-the-box, to make it blocking when no packet is available to be popped.

Reply:
I just rewrote the corresponding part in quic-go because the alloc was showing up in profiles: https://github.com/quic-go/quic-go/pull/3702

Reply:
I see. Cool. That's indeed a solution that does work. Will look into implementing this tomorrow.

Reply:
This is now implemented, with some delay in when I found time to implement it due to personal reasons.
Thanks for the tip!